app/upgrades.go

```diff
@@ -198,3 +200,29 @@ func v010400rc6UpgradeHandler(app *Quicksilver) upgradetypes.UpgradeHandler {
 		return app.mm.RunMigrations(ctx, app.configurator, fromVM)
 	}
 }
+
+func v010400rc8UpgradeHandler(app *Quicksilver) upgradetypes.UpgradeHandler {
+	return func(ctx sdk.Context, _ upgradetypes.Plan, fromVM module.VersionMap) (module.VersionMap, error) {
+		// remove expired failed redelegation records
+		app.InterchainstakingKeeper.IterateZones(ctx, func(index int64, zone icstypes.Zone) (stop bool) {
+			app.InterchainstakingKeeper.IterateAllDelegations(ctx, &zone, func(delegation icstypes.Delegation) (stop bool) {
+				if delegation.RedelegationEnd < 0 {
+					delegation.RedelegationEnd = 0
+					app.InterchainstakingKeeper.SetDelegation(ctx, &zone, delegation)
+				}
+				return false
+			})
+			return false
+		})
+
+		app.InterchainstakingKeeper.IterateRedelegationRecords(ctx, func(_ int64, key []byte, record icstypes.RedelegationRecord) (stop bool) {
+			if record.CompletionTime.Unix() <= 0 {
+				app.InterchainstakingKeeper.Logger(ctx).Info("Removing delegation record", "chainid", record.ChainId, "source", record.Source, "destination", record.Destination, "epoch", record.EpochNumber)
+				app.InterchainstakingKeeper.DeleteRedelegationRecord(ctx, record.ChainId, record.Source, record.Destination, record.EpochNumber)
```

Comment:
Currently not testing this branch of the `if`.  Otherwise, lgtm with @joe-bowman's comments addressed.

Reply:
actually it does the test L153:uprades_test.go - I resued the earlier state to test ; )